fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/matchers/AnyOfMatcher.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * AllOfMatcher.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.matchers;
+
+import com.apple.foundationdb.annotation.API;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Stream;
+
+/**
+ * A matcher that matches if any of its downstream matchers produce bindings. This matcher is
+ * intended to be used to express a logical <em>or</em> between matchers on an object.
+ *
+ * As an example
+ *
+ * {@code
+ * matchinAllOf(greaterThan(0), divisibleBy(2)).matches(2))
+ * }
+ *
+ * produces a stream of one binding which binds to {@code 2}.
+ *
+ * @param <T> the type that this matcher binds to
+ */
+@API(API.Status.EXPERIMENTAL)
+public class AnyOfMatcher<T> implements BindingMatcher<T> {
+    private final Class<T> staticClassOfT;
+    private final List<BindingMatcher<?>> extractingMatchers;
+
+    private AnyOfMatcher(@Nonnull final Class<T> staticClassOfT, @Nonnull final Collection<? extends BindingMatcher<?>> matchingExtractors) {
+        this.staticClassOfT = staticClassOfT;
+        this.extractingMatchers = ImmutableList.copyOf(matchingExtractors);
+    }
+
+    @Nonnull
+    @Override
+    public Class<T> getRootClass() {
+        return staticClassOfT;
+    }
+
+    /**
+     * Attempts to match this matcher against the given object.
+     *
+     * @param outerBindings preexisting bindings to be used by the matcher
+     * @param in the bindable we attempt to match
+     * @return a stream of {@link PlannerBindings} containing the matched bindings, or an empty stream is no match was found
+     */
+    @Nonnull
+    @Override
+    public Stream<PlannerBindings> bindMatchesSafely(@Nonnull PlannerBindings outerBindings, @Nonnull T in) {
+        return extractingMatchers
+                .stream()
+                .flatMap(extractingMatcher -> extractingMatcher.bindMatches(outerBindings, in));
```

Comment:
That is a correct observation and intended. The important detail is that the matching process always binds exactly one object per `PlannerBindings`. If there are more than one match, they get _multiplied_ out. This if you have more than one object matching here you will get a multitude of `PlannerBinding`s on top level, which then causes i.e. the `CascadesPlanner` to call the transform rule for every matching e.g. sub plan.

Example: Let's say you look for an index scan underneath a filter plan. It can be that among the variations we have in the memo structure there is more than one variant that first the bill. In that case you want to call the same rule multiple times. Each time a different index plan is bound to the matcher for index plans.